crates/fast_compute/src/layer.rs

```diff
@@ -0,0 +1,506 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::{cell::RefCell, iter::repeat_with, marker::PhantomData, mem::MaybeUninit, slice};
+
+use binius_compute::{
+	alloc::{BumpAllocator, ComputeAllocator},
+	cpu::layer::count_total_local_buffer_sizes,
+	layer::{
+		ComputeLayer, Error, FSlice, FSliceMut, KernelBuffer, KernelMemMap, KernelMemMapChunk,
+	},
+	memory::{ComputeMemory, SizedSlice, SlicesBatch, SubfieldSlice},
+};
+use binius_field::{
+	ExtensionField, PackedExtension, PackedField,
+	tower::{PackedTop, TowerFamily},
+	unpack_if_possible, unpack_if_possible_mut,
+	util::inner_product_par,
+};
+use binius_math::{
+	ArithExpr, CompositionPoly, RowsBatchRef, fold_left, fold_right, tensor_prod_eq_ind,
+};
+use binius_maybe_rayon::{iter::ParallelIterator, slice::ParallelSliceMut};
+use binius_ntt::AdditiveNTT;
+use binius_utils::{
+	checked_arithmetics::{checked_int_div, strict_log_2},
+	rayon::get_log_max_threads,
+};
+use bytemuck::zeroed_vec;
+use itertools::izip;
+
+use crate::{
+	arith_circuit::ArithCircuitPoly,
+	fri::fold_interleaved_allocated,
+	memory::{PackedMemory, PackedMemorySliceMut},
+};
+
+#[derive(Debug)]
+pub struct FastCpuExecutor;
+
+/// Optimized CPU implementation of the compute layer.
+#[derive(Debug, Default)]
+pub struct FastCpuLayer<T: TowerFamily, P: PackedTop<T>>(PhantomData<(T, P)>);
+
+impl<T: TowerFamily, P: PackedTop<T>> ComputeLayer<T::B128> for FastCpuLayer<T, P> {
+	type Exec = FastCpuExecutor;
+	type KernelExec = FastCpuExecutor;
+	type DevMem = PackedMemory<P>;
+	type OpValue = T::B128;
+	type KernelValue = T::B128;
+	type ExprEval = ArithCircuitPoly<T::B128>;
+
+	fn host_alloc(&self, n: usize) -> impl AsMut<[T::B128]> + '_ {
+		zeroed_vec(n)
+	}
+
+	fn copy_h2d(
+		&self,
+		src: &[T::B128],
+		dst: &mut FSliceMut<'_, T::B128, Self>,
+	) -> Result<(), Error> {
+		debug_assert_eq!(
+			src.len(),
+			dst.len(),
+			"precondition: src and dst buffers must have the same length"
+		);
+
+		unpack_if_possible_mut(
+			dst.data,
+			|scalars| {
+				scalars.copy_from_slice(src);
+				Ok(())
+			},
+			|packed| {
+				let mut iter = src.iter().copied();
+				for p in packed {
+					*p = PackedField::from_scalars(&mut iter)
+				}
+				Ok(())
+			},
+		)
+	}
+
+	fn copy_d2h(&self, src: FSlice<'_, T::B128, Self>, dst: &mut [T::B128]) -> Result<(), Error> {
+		debug_assert_eq!(
+			src.len(),
+			dst.len(),
+			"precondition: src and dst buffers must have the same length"
+		);
+
+		let dst = RefCell::new(dst);
+		unpack_if_possible(
```

Comment:
I don't see why we'd need to support non-PackedFieldIndexable fields. Can we simplify this by making that a requirement?

Reply:
What if we ever decide to use byte-sliced packed fields?

Reply:
Ah, of course. Yes, we can keep this approach, thanks.